test/unit/modules/blocks/process.js

```diff
@@ -15,462 +15,2332 @@
+						expect(cb).to.be.undefined;
+						done();
 					}
 				);
 			});
 
-			describe('when rpc.commonBlock call returns valid result', () => {
-				before(() => {
-					validCommonBlock = Object.assign({}, blockHeightTwo);
-				});
+			it('should return when timestamps are equals and block.id > lastBlock.id', done => {
+				__private.receiveForkOne(
+					{ timestamp: 1, id: 2 },
+					{ timestamp: 1, id: 1 },
+					(err, cb) => {
+						expect(err).to.be.undefined;
+						expect(cb).to.be.undefined;
+						done();
+					}
+				);
+			});
+		});
 
-				it('should return error = null', () => {
-					expect(commonBlockValidationError).to.be.undefined;
-				});
+		describe('Last block and parent loses', () => {
+			beforeEach(() => {
+				__private.validateBlockSlot = sinonSandbox.stub();
 			});
 
-			describe('when rpc.commonBlock call returns invalid result', () => {
-				describe('when id = null', () => {
-					before(() => {
-						validCommonBlock = Object.assign({}, blockHeightTwo);
-						validCommonBlock.id = null;
-					});
+			afterEach(() => {
+				expect(loggerStub.info.args[0][0]).to.equal(
+					'Last block and parent loses'
+				);
+				expect(
+					modules.delegates.fork.calledWithExactly(sinonSandbox.match.object, 1)
+				).to.be.true;
+			});
 
-					it('should return array of errors', () => {
-						expect(commonBlockValidationError)
-							.to.be.an('array')
-							.of.length(1);
+			describe('library.logic.block.objectNormalize', () => {
+				describe('when fails', () => {
+					beforeEach(() => {
+						library.logic.block.objectNormalize.throws('objectNormalize-ERR');
 					});
 
-					it('should return error containing message', () => {
-						expect(commonBlockValidationError)
-							.to.have.nested.property('0.message')
-							.equal('Expected type string but found type null');
+					it('should throw error', done => {
+						__private.receiveForkOne(
+							{ timestamp: 1, id: 2 },
+							{ timestamp: 2, id: 1 },
+							(err, cb) => {
+								expect(err.name).to.equal('objectNormalize-ERR');
+								expect(cb).to.be.undefined;
+								expect(loggerStub.error.args[0][0]).to.equal(
+									'Fork recovery failed'
+								);
+								expect(loggerStub.error.args[0][1].name).to.equal(
+									'objectNormalize-ERR'
+								);
+								done();
+							}
+						);
 					});
+				});
+
+				describe('when succeeds', () => {
+					describe('__private.validateBlockSlot', () => {
+						describe('when fails', () => {
+							beforeEach(() => {
+								library.logic.block.objectNormalize.returns({
+									timestamp: 1,
+									id: 2,
+								});
+								__private.validateBlockSlot.callsArgWith(
+									2,
+									'validateBlockSlot-ERR',
+									null
+								);
+							});
 
-					it('should return error containing path', () => {
-						expect(commonBlockValidationError)
-							.to.have.nested.property('0.path')
-							.equal('#/id');
+							it('should return error', done => {
```

Comment:
it calls a callback with error, it's async - globally please :)

Reply:
done all